### PR TITLE
Issues/56/email integration

### LINKED ISSUE #45: Implement re-login after user credentials change after profile update
When updating non-username or password params on production (heroku) build, the API returns "200" status but does not update the user profile in the db

### LINKED ISSUE #56: Email Verification
* Require email verification before being able to add flights or scoresheets
* Use NodeJS with email API tokens in environmental parameters

### LINKED ISSUE #120: User Password Reset
Users should be able to reset their own password via a password reset email

### PR DESCRIPTION
**Pull request by prphntm63**
- fixed #120 
- fixed #45 
- fixed #56 

Other:
- Added dotenv library. You will need to create a .env file for environmental vars now. `example.env` contains all requried vars.
- Added email service. When running in `NODE_ENV=development`, this will "send" emails using [ethereal](https://ethereal.email/). You can preview these virtual emails using the link generated in the server console.
- Added a status message to `index.pug` that will show when users log in informing them they will need to validate their email address

### CODE REVIEW

File: views/register.pug [hunk 1]
```diff
@@ -87,5 +87,14 @@ block content
...
       )
+      .form-check.mb-3
+        input#allow_automated_email.form-check-input(
+          type="checkbox",
+          name="allow_automated_email",
+          value="true",
+          checked
+        )
+        label.form-check-label(for="allow_automated_email") Allow email communication
+        small.form-text.text-muted.mt-0 For competition updates and information. We will <strong>never</strong> send marketing or advertising emails.
```
Comment (PH-Lazurous) [inline views/register.pug:98]: For competition updates and information only. - Just a grammatical thing, not required.

File: views/unsubscribe.pug [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      )
+      .form-check.mb-3
+        input#allow_automated_email.form-check-input(
+          type="checkbox",
+          name="allow_automated_email",
+          value="true",
+          checked
+        )
+        label.form-check-label(for="allow_automated_email") Allow email communication
+        small.form-text.text-muted.mt-0 For competition updates and information. We will <strong>never</strong> send marketing or advertising emails.
```
Comment (PH-Lazurous) [inline views/unsubscribe.pug:25]: For competition updates and information only. - Just a grammatical thing, not required. (repeat)

File: controllers/AuthController.js [hunk 3]
```diff
@@ -186,14 +189,25 @@ userController.doRegister = function (req, res) {
...
       }).then(function (user) {
-        delete user.password;
-
         req.login(user, (err) => {
           if (err) {
             errorProcessor(err, req);
           }
-          req.flash("success", "Registration Successful!");
+
+          // TODO: Send email with verification code
```
Comment (prphntm63) [inline controllers/AuthController.js:201]: Remove comment

File: controllers/AuthController.js [hunk 4]
```diff
@@ -368,4 +390,234 @@ userController.resetPassword = async function (req, res) {
...
+    {
+      verification_id: emailVerificationCode,
+    },
+    {
+      where: {
+        id: userId,
+      },
+    }
+  ).then(() => {
+    // TODO: Send email with verification code
```
Comment (prphntm63) [inline controllers/AuthController.js:408]: Remove comment

File: controllers/AuthController.js [hunk 5]
```diff
@@ -368,4 +390,234 @@ userController.resetPassword = async function (req, res) {
...
+    );
+    res.redirect("/");
+  });
+};
+
+// Called when user clicks link in their email inbox to validate their account
+userController.validateEmail = function (req, res) {
+  const validationCode = req.query.key;
+  const userId = req.user.id;
+
```
Comment (prphntm63) [inline controllers/AuthController.js:423]: Add escape condition if `!validationCode`

**Comment by prphntm63** [comment]
Tested the 60 min timeout for "invalidating" reset codes in the pg staging environment. Works as intended.

File: controllers/AuthController.js [hunk 6]
```diff
@@ -368,4 +390,234 @@ userController.resetPassword = async function (req, res) {
...
+        password_reset_id: passwordKey,
+      },
+      returning: true,
+      plain: true,
+    }
+  )
+    .then((user) => {
+      if (!user[1]) {
+        req.flash("danger", "Could not reset password. Please try again.");
+        res.redirect("/");
```
Comment (prphntm63) [inline controllers/AuthController.js:575]: add `return`

**Pull request merged**